feat: allow Typescript enums

jaulz opened the pull request:
Right now, all generated types are Typescript types but it would be nice to transform PHP enums into Typescript enums. This PR allows developers to output enums but does not break existing behaviour.

As a developer you could use the following transformer to transform Spatie enums:
```php
<?php

namespace App\Support\TypeScript\Transformers;

use ReflectionClass;
use Spatie\Enum\Enum;
use Spatie\TypeScriptTransformer\Structures\TransformedType;
use Spatie\TypeScriptTransformer\Transformers\Transformer;

class SpatieEnumTransformer implements Transformer
{
    public function transform(ReflectionClass $class, string $name): ?TransformedType
    {
        if ($class->isSubclassOf(Enum::class) === false) {
            return null;
        }

        return TransformedType::create(
            $class,
            $name,
            $this->resolveOptions($class),
            null,
            false,
            'enum'
        );
    }

    private function resolveOptions(ReflectionClass $class): string
    {
        /** @var \Spatie\Enum\Enum $enum */
        $enum = $class->getName();

        $options = collect($enum::toArray())->mapWithKeys(function ($value, $key) {
            return [$key => "'{$key}' = '{$value}'"];
        })->toArray();

        return implode(', ', $options);
    }
}
```

jdreesen commented on src/Structures/TransformedType.php:
```diff
@@ -87,4 +92,16 @@ public function replaceSymbol(string $class, string $replacement): void
     }
+
+    public function toString() {
+        switch ($this->format) {
+            case 'enum': {
+                return "enum {$this->name} { {$this->transformed} };";
+            }
+
+            default: {
+                return "type {$this->name} = {$this->transformed};";
+            }
+        }
```
Since this package is >= PHP 8, we can use `match` instead of `switch` here:

```suggestion
        return match ($this->format) {
            'enum' => "enum {$this->name} { {$this->transformed} };",
            default => "type {$this->name} = {$this->transformed};",
        };
```

rubenvanassche commented:
Hi @jaulz,

It looks like the tests are failing, could you take a look at that?

jaulz commented:
Sorry, it seems that I didn't get any notification about this and just saw your comments by coincidence 😊  

I fixed the tests and also added new tests. Though, since I don't have PHP 8.1 locally I just tried to create the native enum test as similar as possible to the other one. One thing I know is that the tests already miss backed Enum types as I couldn't figure out how to retrieve the "js" or "php" value, e.g.:
```php
enum Enum: string
{
    case JS = 'js';
    case PHP = 'php';
}

$enum = (new ReflectionEnum(Enum::class));
foreach ($enum->getCases() as $case) {
    echo var_dump($case->getValue()); // Returns "enum(Enum::JS)" or "enum(Enum::PHP)"
}
```

jaulz commented:
@rubenvanassche @jdreesen do you think these changes are enough?

rubenvanassche commented:
Hu @jaulz, thanks! I've merged your PR and made some changes, I think it would be much better if the config decides if we transform to native enums or not

jaulz commented:
@rubenvanassche absolutely legitimate... didn't want to touch too much of the existing code so the PR would not be rejected immediately 😉

rubenvanassche commented:
Thanks for the great addition!
